### PR TITLE
Betting

### PR DESCRIPTION
**Pull request by WannabeSmith**
Add some of the betting code (from https://arxiv.org/pdf/2010.09686.pdf) as well as associated unit tests.
Moreover, this runs pytest (but ignores the python files in the googletest library).
Note some of these tests are random (e.g. checking for correct coverage) so they may fail with some (rather small) probability.

### CODE REVIEW

File: src/confseq/misc.py
Comment (gostevehoward) [inline src/confseq/misc.py:11]: i suggest at least
```suggestion
def superMG_crossing_fraction(mart_fn, dist_fn, alpha, repeats):
    def replication():
        x = dist_fn()
        mart = mart_fn(x)
        return any(mart > 1 / alpha)
    return np.mean(replication() for _ in range(repeats)
```
it's not really that bad as a one-liner,
```python
def superMG_crossing_fraction(mart_fn, dist_fn, alpha, repeats):
    return np.mean(any(mart_fn(dist_fn()) > 1 / alpha) for _ in range(repeats)
```
but i leave it up to your taste. (i'm not sure if numpy will take the generator expression but if not you can use a list comprehension or just `sum(<generator expression>) / repeats`.)

File: src/confseq/betting.py
Comment (gostevehoward) [inline src/confseq/betting.py:185]: these look out of date, should be "fns" and document that it's allowed to be a list.

**Review by gostevehoward — APPROVED** [review]
Thank you, these are very exciting contributions.

My overall feeling is that `betting.py` has a lot of lines relative to the actual amount of logic / information contained in the code. There's a lot of duplication, not straight-up copy-pasting of code, but conceptual duplication. I think we could reduce that, making the code a lot easier to read and maintain. Here are some ideas:
- there are certain "groups" of parameters that appear over and over. for example, N, convex_comb, theta, trunc_scale, and m_trunc appear for anything that ultimately uses a betting martingale. breaks, running_intersection, and parallel appear for anything that makes a CS. it's often a good idea to wrap these parameter groups into their own dumb object. if nothing else, it saves us a lot of boilerplate code passing the groups on down to the next function. over time, it may emerge that pieces of logic can naturally move into these objects. (perhaps betting_mart and betting_cs will want to move into such objects, and then they'll no longer be dumb. but i'm happy to start with the dumb object and iterate.) another benefit is that you can move documentation for these parameters to a single place, the parameter object, and you won't need to duplicate it. docstrings is code, too, and though duplication is often permissible in docstrings (unlike code), it does increase maintenance overhead and increase the risk of stale documentation, so reducing duplicated documentation would be great.
- a huge amount of code is duplicated because we do almost the same thing for the positive and negative martingales, or for the upper and lower confidence bounds. one way to avoid this is to rely on the fact that an upper martingale for 1 - X (which also has support in [0,1]) is a lower martingale for X, and likewise a UCB U for 1 - mu yields a LCB 1 - U for mu. So you can write core functions that only produce positive martingales and lower confidence bounds, then write high-level wrappers that call the low-level functions twice, once with X and once with 1 - X.
- the running intersection is always done as a sort of last post-processing step, i think. you could write a separate function `running_intersection(l, u)` and leave it to the user to call it if they want it. (if you have it accept a pair instead of two parameters, it can directly wrap the underlying CS function, though i guess you could use * for that too if you wanted.) where possible, this sort of functional design often opens up more possibilities down the road than having bunch of parameters.

If you want to commit as-is, I'm ok with that, but I do think the above could lead to some dramatic shortening / improvement in the code.

**Comment by gostevehoward** [comment]
I realized you have all these PRs stacked on top of this, so changing it before merging would be a hassle. Let's save the above idea for a future PR.

I also realized the idea about 1 - X can't be applied at the CS level because you do use a two-sided (super)martingale. But I think it could still help a lot at the martingale level.

**Comment by WannabeSmith** [comment]
These are all great points. I'm going to try to address them before adding new features, especially the point about duplicated code/docs. Thank you so much for the detailed review!

File: src/confseq/betting.py
Comment (WannabeSmith) [reply src/confseq/betting.py:185]: Fixed in #12

File: src/confseq/misc.py
Comment (WannabeSmith) [reply src/confseq/misc.py:11]: Updated in #12

**Pull request merged**
